scm/scm_manager.go

```diff
@@ -0,0 +1,173 @@
+package scm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/beatlabs/github-auth/app"
+	"github.com/beatlabs/github-auth/key"
+	"github.com/google/go-github/v45/github"
+	"github.com/quickfeed/quickfeed/internal/env"
+	"go.uber.org/zap"
+)
+
+// SCMManager keeps provider-specific configs (currently only for GitHub)
+// and a map of scm clients for each course.
+type SCMManager struct {
+	Scms      *Scms
+	appConfig *app.Config
+}
+
+// SCMConfig stores SCM varibles.
+type SCMConfig struct {
+	AppID        string
+	AppKey       string
+	ClientID     string
+	ClientSecret string
+}
+
+// NewSCMConfig creates a new SCMConfig.
+func NewSCMConfig() (*SCMConfig, error) {
+	appID, err := env.AppID()
+	if err != nil {
+		return nil, err
+	}
+	clientID, err := env.ClientID()
+	if err != nil {
+		return nil, err
+	}
+	clientSecret, err := env.ClientSecret()
+	if err != nil {
+		return nil, err
+	}
+	appKey := env.AppKey()
+	return &SCMConfig{
+		AppID:        appID,
+		AppKey:       appKey,
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+	}, nil
+}
+
+// NewSCMManager creates base client for the QuickFeed GitHub Application.
+// This client can be used to install API clients for each course organization.
+func NewSCMManager(c *SCMConfig) (*SCMManager, error) {
+	createAppKey, err := key.FromFile(c.AppKey)
+	if err != nil {
+		return nil, fmt.Errorf("error reading key from file: %w", err)
+	}
+	appConfig, err := app.NewConfig(c.AppID, createAppKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating GitHub application client: %w", err)
+	}
+	return &SCMManager{
+		appConfig: appConfig,
+		Scms:      NewScms(),
+	}, nil
+}
+
+func (s *SCMManager) SCMWithToken(ctx context.Context, logger *zap.SugaredLogger, organization string) (SCM, error) {
+	scmClient, err := s.GetOrCreateSCM(ctx, logger, organization)
+	if err != nil {
+		return nil, err
+	}
+	token, err := s.getInstallationToken(ctx, organization)
+	if err != nil {
+		return nil, err
+	}
+	scmClient.SetToken(token)
+	return scmClient, err
+}
+
+// SCMClient gets an existing SCM client by organization name or creates a new client for course organization.
+func (s *SCMManager) GetOrCreateSCM(ctx context.Context, logger *zap.SugaredLogger, organization string) (SCM, error) {
```

Comment:
Okay. That makes sense. Perhaps we should name the function `RefreshToken` and add a doc comment to explain this.

Reply:
On the other hand, `SetToken` is only called from this location, so it doesn't need to be exported. Or do you foresee that it will be called from other places as well?

Reply:
No, I think it is only useful for cloning.